python/plugins/smolagent/composio_smol/toolset.py

```diff
@@ -0,0 +1,193 @@
+    # Initialize toolset
+    composio_toolset = ComposioToolSet()
+
+    tools = composio_toolset.get_tools(
+        actions=[Action.GITHUB_STAR_A_REPOSITORY_FOR_THE_AUTHENTICATED_USER],
+    )
+    # Create agent with Composio tools
+    agent = CodeAgent(
+        tools=list(tools),
+        model=HfApiModel()
+    )
+
+    agent.run("Star the composiohq/composio repo")
+
+    ```
+    """
+
+    def _wrap_action(
+        self,
+        action: str,
+        description: str,
+        schema_params: t.Dict,
+        entity_id: t.Optional[str] = None,
+    ):
+        def function(**kwargs: t.Any) -> t.Dict:
+            """Wrapper function for composio action."""
+            self.logger.debug(f"Executing action: {action} with params: {kwargs}")
+            return self.execute_action(
+                action=action,
+                params=kwargs,
+                entity_id=entity_id or self.entity_id,
+                _check_requested_actions=True,
+            )
+
+        action_func = types.FunctionType(
+            function.__code__,
+            globals=globals(),
+            name=action,
+            closure=function.__closure__,
+        )
+        action_func.__signature__ = Signature(  # type: ignore
+            parameters=get_signature_format_from_schema_params(
+                schema_params=schema_params
+            )
+        )
+
+        action_func.__doc__ = description
+
+        return action_func
+
+    def _wrap_tool(
+        self,
+        schema: t.Dict[str, t.Any],
+        entity_id: t.Optional[str] = None,
+    ) -> StructuredTool:
+        """Wraps composio tool as Langchain StructuredTool object."""
+        action = schema["name"]
+        description = schema["description"]
+        schema_params = schema["parameters"]
+        action_func = self._wrap_action(
+            action=action,
+            description=description,
+            schema_params=schema_params,
+            entity_id=entity_id,
+        )
+
+        # Flatten and format the parameters structure
+        params = {}
+        if "properties" in schema_params:
+            required_params = schema_params.get("required", [])
+            for param_name, param_info in schema_params["properties"].items():
+                param_type = param_info.get("type", "string")
+                # Convert array type to match AUTHORIZED_TYPES
+                if param_type == "array":
+                    param_type = "object"
+
+                param_dict = {
+                    "type": param_type,
+                    "description": param_info.get("description", ""),
+                }
+
+                # Parameter is nullable if it has a default value or is not in required list
+                if (
+                    param_info.get("default") is not None
+                    or param_name not in required_params
+                ):
+                    param_dict["nullable"] = True
+
+                params[param_name] = param_dict
```

Comment:
There's already a function for this in the SDK, maybe in `composio.utils.shared`